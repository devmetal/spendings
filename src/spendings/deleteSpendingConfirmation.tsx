import React, { FC } from 'react';
import ListGroup from 'react-bootstrap/ListGroup';

interface DeleteSpendingConfirmationProps {
  onCancel: () => void;
  onDelete: () => void;
  label?: string;
}

export const DeleteSpendingConfirmation: FC<DeleteSpendingConfirmationProps> = ({
  onCancel,
  onDelete,
  label = 'Are you sure?',
}) => (
  <ListGroup.Item>
    <ListGroup horizontal>
      <ListGroup.Item>{label}</ListGroup.Item>
      <ListGroup.Item variant="primary" onClick={onCancel}>
        No
      </ListGroup.Item>
      <ListGroup.Item variant="danger" onClick={onDelete}>
        Yes
      </ListGroup.Item>
    </ListGroup>
  </ListGroup.Item>
);
