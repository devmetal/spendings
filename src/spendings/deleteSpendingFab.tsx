import { FC, useState } from 'react';
import {
  CircularProgress,
  createStyles,
  Fab,
  makeStyles,
  Theme,
} from '@material-ui/core';
import DeleteIcon from '@material-ui/icons/Delete';
import { animated, useSpring } from 'react-spring';
import { red } from '@material-ui/core/colors';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    wrapper: {
      //transition: '1.5s ease-in transform',
      position: 'relative',
      margin: theme.spacing(1),
      zIndex: 1,
    },
    fabProgress: {
      color: red[600],
      position: 'absolute',
      top: -5,
      left: -5,
      zIndex: 1,
    },
  })
);

export interface DeleteSpendingFabProps {
  onDelete: () => void;
}

const AnimatedProgress = animated(
  ({ step, className }: { step: number; className: string }) => (
    <CircularProgress
      size={50}
      variant="determinate"
      className={className}
      value={step}
    />
  )
);

export const DeleteSpendingFab: FC<DeleteSpendingFabProps> = ({ onDelete }) => {
  const classes = useStyles();
  const [down, setIsDown] = useState(false);

  const { step } = useSpring({
    config: { duration: 1700, precision: 0.1 },
    step: down ? 5 : 1,
    immediate: !down,
    onRest: () => {
      if (down) {
        onDelete();
      }
    },
  });

  return (
    <div
      className={classes.wrapper}
      onTouchStart={(e) => {
        e.stopPropagation();
        setIsDown(true);
      }}
      onTouchEnd={(e) => {
        e.stopPropagation();
        setIsDown(false);
      }}
      onMouseDown={(e) => {
        e.stopPropagation();
        setIsDown(true);
      }}
      onMouseUp={(e) => {
        e.stopPropagation();
        setIsDown(false);
      }}
      onMouseOut={() => {
        setIsDown(false);
      }}
      style={down ? { transform: 'scale(3.0)' } : undefined}
    >
      <Fab
        size="small"
        aria-label="delete"
        role="delete-item"
        color="secondary"
      >
        <DeleteIcon />
      </Fab>
      <AnimatedProgress
        className={classes.fabProgress}
        step={step
          .to((s) => Math.floor(s))
          .to({
            range: [1, 2, 3, 4, 5],
            output: [0, 25, 50, 75, 100],
          })}
      />
    </div>
  );
};
