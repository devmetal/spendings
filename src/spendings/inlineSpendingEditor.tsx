import React, { FC } from 'react';
import ListGroup from 'react-bootstrap/ListGroup';
import Button from 'react-bootstrap/Button';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import { TSpending } from '../types';
import { SpendingInput } from './spendingInput';

interface InlineSpendingEditorProps {
  id: number;
  name: string;
  amount: number;
  onUpdate: (spending: TSpending) => void;
  onCancel: () => void;
}

export const InlineSpendingEditor: FC<InlineSpendingEditorProps> = ({
  id,
  name,
  amount,
  onUpdate,
  onCancel,
}) => {
  const handleUpdate = ({ amount, name }: { amount: number; name: string }) =>
    onUpdate({ id, name, amount });

  return (
    <ListGroup.Item>
      <Row>
        <Col>
          <SpendingInput
            id={id}
            name={name}
            amount={amount}
            onSave={handleUpdate}
            disabled={false}
          />
        </Col>
        <Col>
          <Button onClick={onCancel}>Cancel</Button>
        </Col>
      </Row>
    </ListGroup.Item>
  );
};
