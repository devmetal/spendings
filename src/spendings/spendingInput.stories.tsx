import React from 'react';
import { Meta, Story } from '@storybook/react/types-6-0';
import { SpendingInput, SpendingInputProps } from './spendingInput';

export default {
  title: 'Money/SpendingsInput',
  component: SpendingInput,
} as Meta;

const Template: Story<SpendingInputProps> = (args) => (
  <SpendingInput {...args} />
);

export const Basic = Template.bind({});
Basic.args = {};
