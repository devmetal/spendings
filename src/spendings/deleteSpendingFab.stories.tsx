import { Meta, Story } from '@storybook/react/types-6-0';
import { DeleteSpendingFab, DeleteSpendingFabProps } from './deleteSpendingFab';
import { CssBaseline } from '@material-ui/core';

export default {
  title: 'Money/DeleteSpendingFab',
  component: DeleteSpendingFab,
} as Meta;

const Template: Story<DeleteSpendingFabProps> = (args) => (
  <CssBaseline>
    <DeleteSpendingFab {...args} />
  </CssBaseline>
);

export const Basic = Template.bind({});
Basic.args = {};
