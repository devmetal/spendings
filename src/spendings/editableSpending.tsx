import React, { FC } from 'react';
import ListGroup from 'react-bootstrap/ListGroup';
import { TSpending } from '../types';
import { DeleteSpendingConfirmation } from './deleteSpendingConfirmation';
import { InlineSpendingEditor } from './inlineSpendingEditor';

interface EditableSpendingProps {
  id: number;
  name: string;
  amount: number;
  userEditing: boolean;
  userDeleting: boolean;
  onDelete: (id: number) => void;
  onUpdate: (spending: TSpending) => void;
  onStartUpdate: () => void;
  onCancelUpdate: () => void;
  onStartDelete: () => void;
  onCancelDelete: () => void;
}

export const EditableSpending: FC<EditableSpendingProps> = ({
  id,
  name,
  amount,
  onDelete,
  onUpdate,
  onStartUpdate,
  onCancelUpdate,
  onStartDelete,
  onCancelDelete,
  userEditing,
  userDeleting,
}) => {
  if (userEditing) {
    return (
      <InlineSpendingEditor
        id={id}
        name={name}
        amount={amount}
        onCancel={onCancelUpdate}
        onUpdate={onUpdate}
      />
    );
  }

  if (userDeleting) {
    return (
      <DeleteSpendingConfirmation
        onCancel={onCancelDelete}
        onDelete={() => onDelete(id)}
      />
    );
  }

  return (
    <ListGroup.Item>
      <ListGroup horizontal>
        <ListGroup.Item onClick={onStartUpdate}>{name}</ListGroup.Item>
        <ListGroup.Item onClick={onStartUpdate}>{amount}</ListGroup.Item>
        <ListGroup.Item variant="danger" onClick={onStartDelete}>
          Delete
        </ListGroup.Item>
      </ListGroup>
    </ListGroup.Item>
  );
};
