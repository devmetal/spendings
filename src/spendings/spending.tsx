import React, { FC, useState } from 'react';
import { TSpending } from '../types';
import { EditableSpending } from './editableSpending';

export interface SpendingProps {
  id: number;
  name: string;
  amount: number;
  onDelete: (id: number) => void;
  onUpdate: (spending: TSpending) => void;
}

export const Spending: FC<SpendingProps> = ({
  id,
  name,
  amount,
  onDelete,
  onUpdate,
}) => {
  const [editing, setEditing] = useState<boolean>(false);
  const [deleting, setDeleting] = useState<boolean>(false);

  const handleStartUpdate = () => setEditing(true);

  const handleCancelUpdate = () => setEditing(false);

  const handleStartDelete = () => setDeleting(true);

  const handleCancelDelete = () => setDeleting(false);

  const handleUpdate = (spending: TSpending) => {
    setEditing(false);
    onUpdate(spending);
  };

  const handleDelete = (id: number) => {
    setDeleting(false);
    onDelete(id);
  };

  return (
    <EditableSpending
      id={id}
      name={name}
      amount={amount}
      onDelete={handleDelete}
      onUpdate={handleUpdate}
      onStartUpdate={handleStartUpdate}
      onCancelUpdate={handleCancelUpdate}
      onStartDelete={handleStartDelete}
      onCancelDelete={handleCancelDelete}
      userDeleting={deleting}
      userEditing={editing}
    />
  );
};
