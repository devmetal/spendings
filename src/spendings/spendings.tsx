import React, { FC } from 'react';
import ListGroup from 'react-bootstrap/ListGroup';
import { TSpending } from '../types';
import { Spending } from './spending';

export interface SpendingsProps {
  /**
   * List of user spendings
   */
  spendings: Array<TSpending>;
  /**
   * User start update
   */
  onUpdate: (spending: TSpending) => void;
  /**
   * User start delete
   */
  onDelete: (id: number) => void;
}

/**
 * Spendings is a basic list of represent user spendings
 */
export const Spendings: FC<SpendingsProps> = ({
  spendings,
  onUpdate,
  onDelete,
}) => (
  <ListGroup>
    {spendings.map((spending) => (
      <Spending
        key={spending.id}
        id={spending.id}
        name={spending.name}
        amount={spending.amount}
        onUpdate={onUpdate}
        onDelete={onDelete}
      />
    ))}
  </ListGroup>
);
