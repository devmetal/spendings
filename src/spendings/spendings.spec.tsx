import React from 'react';
import { fireEvent, render, screen, wait } from '@testing-library/react';
import { Spendings } from './spendings';

const props = {
  onDelete: jest.fn(),
  onUpdate: jest.fn(),
  spendings: [
    {
      id: 0,
      amount: 10,
      name: 'test',
    },
  ],
};

test('Sanity snapshot testing', () => {
  const { container } = render(<Spendings {...props} />);
  expect(container).toMatchSnapshot();
});

describe('interactions', () => {
  let name: HTMLElement;
  let delBtn: HTMLElement;

  beforeEach(() => {
    const { getByText } = render(<Spendings {...props} />);
    name = getByText('test');
    delBtn = getByText('Delete');
  });

  test('delete interaction should show confirm ui', async () => {
    fireEvent.click(delBtn);
    await wait(() => expect(screen.getByText('Yes')).toBeVisible());
  });

  test('update interaction show update ui', async () => {
    fireEvent.click(name);
    await wait(() => expect(screen.getByText('Cancel')).toBeVisible());
  });
});
