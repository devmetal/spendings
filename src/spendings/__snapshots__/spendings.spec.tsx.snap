// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sanity snapshot testing 1`] = `
<div>
  <div
    class="list-group"
  >
    <div
      class="list-group-item"
    >
      <div
        class="list-group list-group-horizontal"
      >
        <div
          class="list-group-item"
        >
          test
        </div>
        <div
          class="list-group-item"
        >
          10
        </div>
        <div
          class="list-group-item list-group-item-danger"
        >
          Delete
        </div>
      </div>
    </div>
  </div>
</div>
`;
